adding flag option to only pad the last dim

btolooshams opened the pull request:
Added a pad_only_last_dim in DomainPadding class (default False). Updated FNO class to have the flag domain_padding_only_last_dim (default False). If set True, then padding is only applied to the last dim of the input.

JeanKossaifi commented:
Thanks @btolooshams. Let's make it more general, how about having two use cases:
- padding is True, in that case set padding = range(...) such that we pad all dimensions
- Padding is a list of dimensions (for your use case would be [4], in that case we just pad those dimensions.

The only question is whether we account batch-size and channels in the dimensions (dim 0 is then batch size). Otherwise 0 would be the first spatial dim. 
  

What do you think? Could you make the change for the padding and update the docstrings in the models?

btolooshams commented:
I have modified the previous pool request. The "pad_only_last_dim" is now changed to "padding_dim". If True, all dimensions (0,1,2) excluding batch and channel with be padded, if "padding_dim" is a list, then those dimensions in the list will only be padded.

An additional fix --> the padding in F is applied in reverse order, so I have added a line to reverse the padding. Padding was being wrongly applied before.

btolooshams commented:
The current implementation excludes batch and ch dim (0,1) in counting with dimension to pad. 0 is the first spatial dimension.

JeanKossaifi commented:
Thank you @btolooshams, this is great! Could we call it just `padding`? 
Would be great to add a quick test case in the unit-tests for sanity and to make sure it remains correct in the future?

JeanKossaifi commented on neuralop/layers/tests/test_padding.py:
```diff
@@ -3,16 +3,44 @@
     target_shape = list(padded.shape)
-    target_shape[-1] = target_shape[-2] = out_size[mode]
+    for ctr in range(1,3):
+        target_shape[-ctr] = out_size[mode][-ctr]
```
Can we just write this out to make it easier to read with a small inline comment? ALso I don't know what `ctr` means, good to use clear names

btolooshams replied:
"ctr" is changed to "pad_dim", and comment added.

JeanKossaifi commented on neuralop/layers/tests/test_padding.py:
```diff
@@ -3,16 +3,44 @@
+
     data = torch.randn((2, 3, 10, 10))
-    padder = DomainPadding(0.2, mode)
+    padder = DomainPadding(padding_dim, mode)
```
Why do we call it padding_dim here?

btolooshams replied:
padding_dim is renamed to "padding"

JeanKossaifi commented on neuralop/layers/padding.py:
```diff
@@ -52,33 +57,50 @@ def pad(self, x):
+                unpad_list = list()
+                for p in output_pad[::-1]:
+                    if p == 0:
+                        unpad_amount_pos = None
```
Why not leave 0 in this case?

btolooshams replied:
slice gives an error for p=0. So if padding is 0, p is set to None, hence no slicing is done to unpad list.

JeanKossaifi commented on neuralop/layers/padding.py:
```diff
@@ -52,33 +57,50 @@ def pad(self, x):
+                    else:
+                        unpad_amount_pos = p
+                        unpad_amount_neg = -p
+                    unpad_list.append(slice(unpad_amount_pos, unpad_amount_neg, None))
```
I actually don't mind spelling it out but can we call it padding_start and padding_end or something clearer? :)

btolooshams replied:
name modified :)

JeanKossaifi commented on neuralop/layers/padding.py:
```diff
@@ -39,8 +40,12 @@ def pad(self, x):
         if isinstance(self.domain_padding, (float, int)):
             self.domain_padding = [float(self.domain_padding)]*len(resolution)
+
+        assert len(self.domain_padding) == len(resolution)
```
Good check to have! Can you add a message if that's not the case to explain the issue to users?

btolooshams replied:
Explanation is added. "domain_padding length must match the number of spatial/time dimensions (excluding batch, ch)"

JeanKossaifi commented on neuralop/layers/tests/test_padding.py:
```diff
@@ -3,16 +3,44 @@
+    padded = padder.pad(data)
+
+    target_shape = list(padded.shape)
+    for ctr in range(1,4):
```
Same comment as above - ideally tests should be self explanatory and easy to get by just reading the code!

JeanKossaifi reviewed (commented):
Thanks @btolooshams! I like the API like this, I left a few comments, mostly just style.

JeanKossaifi commented:
Thank you @btolooshams, this is great, merging!
